### ISSUE TITLE
The inferred type of 'X' cannot be named without a reference to '@reduxjs/toolkit/node_modules/immer/dist/internal'. This is likely not portable. A type annotation is necessary.

### ISSUE THREAD

**Issue by filiptrplan** [reduxjs/redux-toolkit#1806]
This error shows up in VS Code when using TypeScript.
The slice code: 

```js
import { POSItem } from './../POSService';
import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { RootState } from '../../../app/store';

export interface POSState {
    addedPOSItems: POSItem[],
}


const initialState = {
    addedPOSItems: []
} as POSState;

export const POSSlice = createSlice({
    name: 'pos',
    initialState,
    reducers: {
        setAddedPOSItems: (state, action: PayloadAction<POSItem[]>) => {
            state.addedPOSItems = action.payload;
        }
    }
});

export const { setAddedPOSItems } = POSSlice.actions;
export const selectAddedPOSItems = (state: RootState) => state.pos.addedPOSItems;
export default POSSlice.reducer;
````

The store config
```js
import { configureStore } from '@reduxjs/toolkit';
import POSSlice from '../client/POS/POSService/POSSlice';

const store = configureStore({
    reducer: {
        pos: POSSlice
    },
});

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
export default store;
```

Versions:
```
    "react": "^17.0.2",
    "redux": "^4.1.0",
    "@reduxjs/toolkit": "^1.6.1",
```

Full error message:
```
"The inferred type of 'POSSlice' cannot be named without a reference to '@reduxjs/toolkit/node_modules/immer/dist/internal'. This is likely not portable. A type annotation is necessary."
```

Reinstalling node_modules or clearing the cache didn't help. Do you need any more information?

**Comment by markerikson** [comment]
There was one prior report of this in #1181 . Unfortunately, this really sounds like some kind of a bundling/installation problem, and not something that's actually wrong with the source code.

Given that thousands of people are using RTK with TypeScript without issues, this is _probably_ something very specific to your project setup, and not anything we can investigate and fix ourselves.

If you can manage to put together a project that demonstrates this happening we might be able to take a look, but otherwise I suspect there's nothing we can do here.

**Comment by filiptrplan** [comment]
I'm quite inexperienced in TypeScript. Can you maybe just point me in a direction where I should start troubleshooting?

**Comment by filiptrplan** [comment]
I solved the error by doing the following in my tsconfig.json

This is probably just a workaround but if someone comes across this problem this should work.

```json
{
  "compilerOptions": {
    "declaration": false,
  },
```

**Comment by james-mungai** [comment]
@filiptrplan thanks

**Comment by bfricka** [comment]
TLDR; No clue why this happens, but you can add `immer` as a dev dependency and still compile with declarations.

Follow up on this one. I'm working with two different monorepos, both using RTK, and both using `"declaration": true`. For reasons that should be obvious, simply setting `"declaration": false` is not a solution (although it does work, if you don't have any need to publish types).

Here's the deal for me. One project has this error (let's call it project CRAP), the other project doesn't (call this one CLAP). I have desperately tried to figure out _why_ this error is occurring in CRAP and not in CLAP. Nearly everything I can think of I have normalized between the two:

- Same tsconfig (nearly, see below)
- Same RTK version
- Same TypeScript version
- The one without the error was importing `castDraft` from `immer`, so I tried that, thinking the import might trigger some bail out logic. Didn't do squat.
- Both packages I tested on within the separate monorepos have RTK peer dependency defined as ^1.7
- The various apps within these projects have RTK as a dependency (shouldn't matter, because it's not the apps but the supporting packages that I'm building and publishing using `tsc`).
- None of the packages within either monorepo have `immer` as a dependency or devDependency.

And when I say the versions are the same, these are both running yarn workspaces, properly deduped. So I do mean the same.

The only thing that's different, that I can see at least, is the project structure and the path mappings in the tsconfigs. In CLAP, the apps themselves are not part of the path mappings (although, everything is still defined in workspaces). Only the "shared" and externally published packages are defined in the path mappings.

I frankly can't see how that would matter though. I also tried, changing the path mappings in CRAP so only the "shared" modules are mapped, and this did nothing.

All of this said, long before I tried any of this, I did find a way to "trick" TypeScript into thinking all is right in the world by adding `immer` as a dev dependency. So I guess that's what I have to do? I know this "fixes" it and I can now compile with declarations and without errors, but it seems so arbitrary.

If I'm able to turn CRAP into CLAP, then I'll update this.

**Comment by ctbwx** [comment]
import { Slice } from '@reduxjs/toolkit';

export const POSSlice: Slice = createSlice({
    name: 'pos',
    initialState,
    reducers: {
        setAddedPOSItems: (state, action: PayloadAction<POSItem[]>) => {
            state.addedPOSItems = action.payload;
        }
    }
});

**Comment by bfricka** [comment]
@ctbwx I have no idea what you're trying to say here, but explicitly typing `Slice` is not how you should ever do this. Like, I can't emphasize this enough. You lose all the type inference of `Name`, `State`, `CaseReducers` (and all the actions).

Maybe you explain how this relates to the issue here, at least.

**Comment by NoahAndrews** [comment]
For me, it only works if I use version `1.5.1` or earlier _and_ I declare the precise version of the dependency in `package.json` (no caret in the version string). Like @bfricka, I'm using Yarn workspaces.

**Comment by PeiLiao** [comment]
`export const templateSlice: any = createSlice({
    name: 'template',
    initialState, *****`

only need to declare the slice explicitly  to resolve this ts type error

**Comment by markerikson** [comment]
@PeiLiao : please don't do that :( You're throwing away all the actual TS types in `templateSlice`.

**Comment by NickBolles** [comment]
Since google brought me to this one first, linking to my comment confirming a fix in the referenced issue: https://github.com/reduxjs/redux-toolkit/issues/1181#issuecomment-1183711921

**Comment by JCB-K** [comment]
Ran into this issue too - fiddling around with Immer didn't work for me, but it resolved itself once all applications in my monorepo were on the exact same (newest) version of redux-toolkit.

**Comment by yuji-luigi** [comment]
> There was one prior report of this in #1181 . Unfortunately, this really sounds like some kind of a bundling/installation problem, and not something that's actually wrong with the source code.
> 
> Given that thousands of people are using RTK with TypeScript without issues, this is _probably_ something very specific to your project setup, and not anything we can investigate and fix ourselves.
> 
> If you can manage to put together a project that demonstrates this happening we might be able to take a look, but otherwise I suspect there's nothing we can do here.

After this post.
I deleted node_modules and pnpm-lock.yaml file and updated all the dependencies. Then the error was gone.
Thank you
Hope it helps others!

**Comment by taylorkline** [comment]
Adding `immer` to devDependencies is not strictly necessary. 

A simple `import "immer"` at the top of the file quiets the error.

**Comment by mmncit** [comment]
> Hey the solution is to:
> 
> * add `immer` to devDependencies
> * add `import 'immer'` on top of your file, this will not import anything at runtime but it will import the types to make the compiler happy
> 
> We could fix this inside `redux-library` by enforcing the type of `slice.actions` to not include `WritableDraft` as part of it (it's an implementation detail either way)

In my case, unfortunately, none of these setups is working.

Attempts:
 
1. not adding `immer` to devDependencies and `import 'immer'` on top of slice.ts file
2. adding `immer` to devDependencies and `import 'immer'` on top of slice.ts file

The only setup works when I disable `declaration` in `compilerOptions`. However, can't do it as the library is exportable. Are there any other settings I might miss, adding `immer`. Is it possible to share a git link of your minimal?

**Comment by ly3xqhl8g9** [comment]
> > Hey the solution is to:
> > 
> > * add `immer` to devDependencies
> > * add `import 'immer'` on top of your file, this will not import anything at runtime but it will import the types to make the compiler happy
> > 
> > We could fix this inside `redux-library` by enforcing the type of `slice.actions` to not include `WritableDraft` as part of it (it's an implementation detail either way)
> 
> In my case, unfortunately, none of these setups is working.
> 
> Attempts:
> 
> 1. not adding `immer` to devDependencies and `import 'immer'` on top of slice.ts file
> 2. adding `immer` to devDependencies and `import 'immer'` on top of slice.ts file
> 
> The only setup works when I disable `declaration` in `compilerOptions`. However, can't do it as the library is exportable. Are there any other settings I might miss, adding `immer`. Is it possible to share a git link of your minimal?

A solution might be adding `"preserveSymlinks": true` in the `"compilerOptions"` of `tsconfig.json`.

**Comment by srosato** [comment]
```
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import type { WritableDraft } from '@reduxjs/toolkit/node_modules/immer/dist/internal';
```

As another potential not the prettiest workaround.

**Comment by nouman91** [comment]
Thank you @srosato this worked!

**Comment by carneiror** [comment]
I got the same error while trying the test `@reduxjs/toolkit@2.0.0-beta.0`. I'm currently working on a big monorepo and the solution was to check versions and syncpack them:

```
$ yarn why redux
$ yarn why immer
```

The issue was that yarn does package hosting ending up with conflicting `.d.ts` files. Depending on the package version that is hoisted to the project root and your tsconfig configuration, you might end up with this type of errors.

Moving all versions to `redux@5.0.0-beta.0` and `immer@10.0.2` fixed this issue for me.

**Comment by Visual-Dawg** [comment]
For me, none of the above works. 
pnpm monorepo with `immer` installed separately too.

**Comment by srosato** [comment]
> For me, none of the above works. pnpm monorepo with `immer` installed separately too.
> 
> Edit: Installing the same `immer` version as `Redux toolkit` uses, in this case it was `immer@9.0.21`, _and_ adding `import "immer"` to the top of the file fixed it.

I also recently switched to `pnpm` and then I did the same as you except that I did not have to use `import 'immer'`

What I did:

1. Get version of immer used by redux-toolkit `cat node_modules/@reduxjs/toolkit/package.json | grep immer`
2. `pnpm i immer@^9.0.21` // in my case redux-toolkit reported using ^9.0.21

Then the error went away. No need to my earlier workaround I posted earlier here in this thread when using `yarn`

**Comment by srosato** [comment]
Actually, I am not sure if using `import 'immer'` might be too expensive of an import, so I tried my earlier workaround and modified it with:

```typescript
/* eslint-disable-next-line @typescript-eslint/no-unused-vars,unused-imports/no-unused-imports */ // noinspection ES6UnusedImports
import type { Draft } from 'immer'
```

The error also goes away like that.

**Comment by SharmaTushar** [comment]
Importing immer didn't work for me, so I referenced it as
```
/// <reference path="../node_modules/immer/dist/immer.d.ts" />
```

**Comment by zkWildfire** [comment]
I came across this same error when using `createEntityAdapter()`. The issue is the type of `state` in the reducer. Instead of letting the type be inferred to be a `WritableDraft`, you need to explicitly specify it. Based on how I fixed it in my code, I believe the code in the OP can be fixed as follows:
```ts
export interface POSState {
    addedPOSItems: POSItem[],
}


const initialState = {
    addedPOSItems: []
} as POSState;

export const POSSlice = createSlice({
    name: 'pos',
    initialState,
    reducers: {
        // Change `state` to `state: POSState`
        //                 vvvvvvvvvvvvvvv
        setAddedPOSItems: (state: POSState, action: PayloadAction<POSItem[]>) => {
            state.addedPOSItems = action.payload;
        }
    }
});
```
In my case, I had code that looks like this:
```ts
const fooAdapter = createEntityAdapter<FooModel>();
export const fooSlice = createSlice({
    name: "foo",
    initialState: fooAdapter.getInitialState({
        myValue: 0
    }),
    reducers: {
        addFoo: FooAdapter.addOne,
        setFoo: FooAdapter.setOne,
        updateFoo: FooAdapter.updateOne,
        removeFoo: FooAdapter.removeOne,
        // Adding this reducer causes the error
        setValue: (state, action: PayloadAction<number>) => {
            state.myValue = action.payload;
        }
    },
});
```
I fixed the error by adding an interface that declared the shape of the state object returned by `fooAdapter` and explicitly adding the type to `state`:
```ts
export interface FooState extends EntityState<FooModel> {
    myValue: number;
}

const fooAdapter = createEntityAdapter<FooModel>();
export const fooSlice = createSlice({
    name: "foo",
    initialState: fooAdapter.getInitialState({
        myValue: 0
    }),
    reducers: {
        addFoo: FooAdapter.addOne,
        setFoo: FooAdapter.setOne,
        updateFoo: FooAdapter.updateOne,
        removeFoo: FooAdapter.removeOne,
        setValue: (state: FooState, action: PayloadAction<number>) => {
            state.myValue = action.payload;
        }
    },
});
```

**Comment by haozi** [comment]
This is how I solved it: 
Add the same version of immer that is used in [@reduxjs/toolkit]( https://unpkg.com/browse/@reduxjs/toolkit/package.json) as a dependency in my project, currently it's `^9.0.21`

Then in ther xxx.slice.ts files top add
```
 /// <reference types="immer" />
```

**Comment by sallyzoughaib** [comment]
@srosato 
i have the same error: The inferred type of 'keysApi' cannot be named without a reference to '@libs/node_modules/axios'. This is likely not portable. A type annotation is necessary.

But I tried  to install Axios locally in a specific folder within my Node.js TypeScript project and import it from that directory, I  followed  these steps: 
1. Create a folder in my  project where I want to install Axios named by ' libs '  
2.  Run npm install axios --prefix ./libs .

![image](https://github.com/reduxjs/redux-toolkit/assets/122433398/4da3077c-9fb7-4de6-b5e4-f9a91ce0869e)

![image](https://github.com/reduxjs/redux-toolkit/assets/122433398/e53a6762-8f85-48f1-be6b-6ac9ca25dc8f)

**Comment by shailesh17** [comment]
The fix for me was to add declaring the slice as `: Slice` in the line where `createSlice` is called:

```
import { createSlice, Slice } from '@reduxjs/toolkit'

type MyState = {
   name: string,
   address: string,
}

const mySlice: Slice = createSlice({
    name: 'mystate',
    initialState,  
    reducers: {
      ...
    },
}
```

**Comment by EskiMojo14** [comment]
@shailesh17 this isn't really a solution as you lose all specific information about the slice - what actions it has, its name, etc

**Comment by skurgansky-sugarcrm** [comment]
this problem is described here https://typescript.tv/errors/#ts2742

**Comment by sam-mfb** [comment]
> Hey the solution is to:
> 
> * add `immer` to devDependencies
> * add `import 'immer'` on top of your file, this will not import anything at runtime but it will import the types to make the compiler happy
> 
> We could fix this inside `redux-library` by enforcing the type of `slice.actions` to not include `WritableDraft` as part of it (it's an implementation detail either way)

We see this issue a lot using rtk in our monorepo and solve it in the way noted here, i.e., an import of immer as a dev dependency.  However, I do think the error is actually a typing error (albeit a minor one) namely that the return type of `createSlice` is dependent on the type of `WriteableDraft` from immer and yet that type is not exposed by rtk.  One way to conceptualize the problem is to think about what would happen if you had a library that relied on rtk and, for some reason, your library wanted to export for its consumers the return value of `createSlice`.  `tsc` wouldn't be able to create the declaration file for that export because it can't "describe" the `WriteableDraft `type based solely on the type exports of rtk. I believe the reason this doesn't show up in many installations is because many folks are using rtk at the "top" level of their app and so the compiler never needs to worry about generating rtk-dependent types--rtk's typing is solely used to compile code, not generate declarations.  but in a monorepo structure it's much easier to end up with internal libraries that have exports (and therefore generate declaration files) that depend on rtk's types, and so these sorts of incomplete type-dependency tree issues get exposed.

It would seem to me the "correct" solution would be for rtk to either re-export the `WriteableDraft` type so the compiler can "see" it when constructing the `createSlice` return type (and any dependent declarations) or alter the typing of createSlice so its return type converts `WriteableDraft<T>` to just T wherever it occurs.  I would think that would be fine as immer is only actually used inside the function.

If it was my library, I'd probably opt for the latter solution.  (And this is, in effect, what the solutions here are doing that suggest typing `state` in the case reducers as--they are effectively replacing `WriteableDraft<State>` with `State`.) 

But, in any event, do @markerikson or @phryneas have a view on whether they'd be willing to implement either of these solutions?  Or an explanation of why i'm off the mark on that this is a problem (however small)?

In fairness the import immer solution as a dev dependency isn't terrible, and as heavy monorepo users we are used to seeing this kind of implicit type dependency when we use third party libs, but rtk is so well typed in general, it always bugs me that it has this little bit of type pollution.  😃

**Comment by EskiMojo14** [comment]
we don't export the WriteableDraft type because we export the [Draft](https://github.com/immerjs/immer/blob/main/src/types/types-external.ts#L36) type.

It also wouldn't be correct for the case reducers to change Draft<T> to T - the point is that Draft<T> makes a readonly object mutable, and by changing that you would be able to unsafely pass a readonly state to a mutating case reducer.
<img width="547" alt="image" src="https://github.com/reduxjs/redux-toolkit/assets/18308300/882c6653-838b-4537-ad37-53543aed3740">

**Comment by boan-anbo** [comment]
> we don't export the WriteableDraft type because we export the [Draft](https://github.com/immerjs/immer/blob/main/src/types/types-external.ts#L36) type.
> 
> It also wouldn't be correct for the case reducers to change Draft to T - the point is that Draft makes a readonly object mutable, and by changing that you would be able to unsafely pass a readonly state to a mutating case reducer. <img alt="image" width="547" src="https://private-user-images.githubusercontent.com/18308300/302068643-882c6653-838b-4537-ad37-53543aed3740.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg0MTkwMzUsIm5iZiI6MTcwODQxODczNSwicGF0aCI6Ii8xODMwODMwMC8zMDIwNjg2NDMtODgyYzY2NTMtODM4Yi00NTM3LWFkMzctNTM1NDNhZWQzNzQwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIwVDA4NDUzNVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTNiOGVmZmNkNjNiOTQ2OTc3YmJjZTNlOWEwYTY3MWU2YTMxMzgwNDM4MzJjM2ExYzQ5YWMyZGUwMTMxMjAxMjYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.eFusRFU4tF1unv3VONHiy2YDUQ0g7GFhXapOryGyrOo">

Thanks! Finally found the alternative to WritableDraft after migrating to RTK 2

**Comment by Dan503** [comment]
None of the solutions mentions above worked for me so I ended up doing this as work around (yes there is an `any` type in there, I don't care, it's close enough).

```js
import {
    AsyncThunk,
    createAsyncThunk,
} from '@reduxjs/toolkit'


export const fetchPosts: AsyncThunk<Array<Post>, void, any> = createAsyncThunk<Array<Post>>(
    'posts/fetchPosts',
    async () => {
	const response = await client.get('/api/posts')
	return response.data
    }
)

```

**Comment by markerikson** [comment]
@Dan503 on a related note, I have an almost-complete rewrite of the "Essentials" tutorial that uses TS here:

- https://github.com/reduxjs/redux/pull/4706

I'd appreciate any feedback you have on that (over in the PR), especially given that you've just gone through the tutorial.

(Also, fwiw, I didn't run into that error myself while updating the example app.)

**Comment by Tenekedzhiev** [comment]
None of the above worked for me either. Here's an alternative:

`tsconfig.json`
```javascript
{
  "compilerOptions": {
    "baseUrl": ".",
    "paths": { "@reduxjs/toolkit": ["node_modules/@reduxjs/toolkit/"] },
   // ...
  }
}

**Comment by aryaemami59** [comment]
@Tenekedzhiev

Can you run this command and try again to see if it fixes the issue:

```bash
npm install https://pkg.csb.dev/reduxjs/redux-toolkit/commit/06a30ee4/@reduxjs/toolkit
```

**Comment by Born2shine** [comment]
I had the same issue but when I run this comman `npm install https://pkg.csb.dev/reduxjs/redux-toolkit/commit/06a30ee4/@reduxjs/toolkit` by @aryaemami59 the issue got fixed

**Comment by Dan503** [comment]
@TomasDannunzioFIT
> I had this issue and to solve it I just had to add a type to the reducer I want to export. In the presented case, it should be: 
>
> `export const POSSlice : any = createSlice({ ... })` 

Please don't do that. You have destroyed all of the types information from the `createSlice` function.

You are no longer getting any benefits from TypeScript by typing `POSSlice` as `any`.

**Comment by markerikson** [comment]
@TomasDannunzioFIT : no, you absolutely **_must not_ do that!!!!**.  As said, that will remove _all_ TS types and destroy any usefulness.

**Comment by markerikson** [comment]
Fixed in https://github.com/reduxjs/redux-toolkit/releases/tag/v2.2.7 !
